zilliqa/src/consensus.rs

```diff
@@ -384,28 +421,127 @@ impl Consensus {
+            .unwrap()
+            .unwrap()
+    }
+
+    // This function is called when we suspect that we are out of sync with the network/need to catchup
+    pub fn download_blocks_up_to_head(&mut self) -> Result<()> {
+        let head_block = self.head_block();
+        self.block_store.request_blocks(head_block.header.number)?;
 
         Ok(())
     }
 
-    pub fn timeout(&mut self) -> Result<(PeerId, NewView)> {
-        self.view += 1;
+    // Set view only here - that way the last_timeout should be set iff view is set
+    fn set_view(&mut self, view: u64) {
+        match view.cmp(&self.view) {
+            std::cmp::Ordering::Less => {
+                warn!(
+                    "Tried to set view {} to lower view {} - this is incorrect",
+                    self.view, view
+                );
+            }
+            std::cmp::Ordering::Equal => {
+                warn!("Tried to set view to same view - this is incorrect");
+            }
+            std::cmp::Ordering::Greater => {
+                self.view = view;
+                self.last_timeout = SystemTime::now();
+            }
+        }
+    }
 
-        let leader = self.get_leader(self.view)?.peer_id;
+    pub fn timeout(&mut self) -> Option<(PeerId, ExternalMessage)> {
+        // We never want to timeout while on view 1
+        if self.view == 1 {
+            let genesis = self
+                .get_block_by_view(0)
+                .unwrap()
+                .ok_or_else(|| anyhow!("missing block"))
+                .unwrap();
+            // If we're in the genesis committee, vote again.
+            if genesis
+                .committee
+                .iter()
+                .any(|v| v.peer_id == self.peer_id())
+            {
+                info!("timeout in view 1, we will vote for genesis block rather than incrementing view");
+                let leader = genesis.committee.leader(self.view);
+                let vote = self.vote_from_block(&genesis);
+                return Some((leader.peer_id, ExternalMessage::Vote(vote)));
+            } else {
+                info!("We are on view 1 but we are not a validator, so we are waiting.");
+                let _ = self.download_blocks_up_to_head();
+            }
+
+            return None;
+        }
+
+        // Now consider whether we want to timeout - the timeout duration doubles every time, so it
+        // Should eventually have all nodes on the same view
+        let consensus_timeout_ms = self.config.consensus.consensus_timeout.as_millis() as u64;
+        let head_block = self.head_block();
+        let head_block_view = head_block.view();
+        trace!("Time since last change is: {:?}", self.last_timeout);
+        println!("Time since last change is: {:?}", self.last_timeout);
+        let time_since_last_view_change = SystemTime::now()
+            .duration_since(self.last_timeout)
+            .expect("last timeout seems to be in the future...")
```

Comment:
Yes, great suggestions - how about `time_since_last_view_change` just defaults to 0s with a warn! if it fails? That way we don't touch/potentially race on `last_timeout`

Reply:
My thinking was that if you adjust the system clock it might invalidate `last_timeout`. Say a machine was running ahead by 5-10 minutes (e.g. NTP misconfigured/not running/whatever), then the system clock gets abruptly fixed to the correct time: you would then wait 5-10 minutes for your next timeout. Whereas if you reset `last_timeout`, you will only wait up to, at the very most, double the normal timeout length.

However machines with such poorly configured system time probably can't properly run the chain anyway and don't need to be supported, so on second thought your more conservative solution of doing nothing (defaulting `time_since_last_view_change` to 0 and hoping the problem goes away eventually) might be just fine.